Fix covery issues from Controller Input PR

garbear opened the pull request:
Coverty also reported the following issues that I can't find a fix for:

MY CODE (controller input PR):

```
________________________________________________________________________________________________________
*** CID 1355590:  Performance inefficiencies  (PASS_BY_VALUE)
/xbmc/games/controllers/Controller.cpp: 32 in GAME::CController::FromExtension(ADDON::AddonProps, const cp_extension_t *)()
26     #include "utils/XBMCTinyXML.h"
27
28     using namespace GAME;
29
30     const ControllerPtr CController::EmptyPtr;
31
>>>     CID 1355590:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter props of type "ADDON::AddonProps" (size 352 bytes) by value.
32     std::unique_ptr<CController> CController::FromExtension(ADDON::AddonProps props, const cp_extension_t* ext)
33     {
34       return std::unique_ptr<CController>(new CController(std::move(props)));
35     }
36
37     CController::CController(ADDON::AddonProps addonprops) :


________________________________________________________________________________________________________
*** CID 1355589:  Performance inefficiencies  (PASS_BY_VALUE)
/xbmc/games/controllers/Controller.cpp: 37 in GAME::CController::CController(ADDON::AddonProps)()
31
32     std::unique_ptr<CController> CController::FromExtension(ADDON::AddonProps props, const cp_extension_t* ext)
33     {
34       return std::unique_ptr<CController>(new CController(std::move(props)));
35     }
36
>>>     CID 1355589:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter addonprops of type "ADDON::AddonProps" (size 352 bytes) by value.
37     CController::CController(ADDON::AddonProps addonprops) :
38       CAddon(std::move(addonprops)),
39       m_bLoaded(false)
40     {
41     }
42


________________________________________________________________________________________________________
*** CID 1355588:  Performance inefficiencies  (PASS_BY_VALUE)
/xbmc/peripherals/addons/PeripheralAddon.cpp: 50 in PERIPHERALS::CPeripheralAddon::FromExtension(ADDON::AddonProps, const cp_extension_t *)()
44     using namespace XFILE;
45
46     #ifndef SAFE_DELETE
47       #define SAFE_DELETE(p)  do { delete (p); (p) = NULL; } while (0)
48     #endif
49
>>>     CID 1355588:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter props of type "ADDON::AddonProps" (size 352 bytes) by value.
50     std::unique_ptr<CPeripheralAddon> CPeripheralAddon::FromExtension(ADDON::AddonProps props, const cp_extension_t* ext)
51     {
52       std::string strProvidesJoysticks = ADDON::CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@provides_joysticks");
53       std::string strProvidesButtonMaps = ADDON::CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@provides_buttonmaps");
54
55       bool bProvidesJoysticks = StringUtils::EqualsNoCase(strProvidesJoysticks, "true");


________________________________________________________________________________________________________
*** CID 1355587:  Performance inefficiencies  (PASS_BY_VALUE)
/xbmc/peripherals/addons/PeripheralAddon.cpp: 61 in PERIPHERALS::CPeripheralAddon::CPeripheralAddon(ADDON::AddonProps, bool, bool)()
55       bool bProvidesJoysticks = StringUtils::EqualsNoCase(strProvidesJoysticks, "true");
56       bool bProvidesButtonMaps = StringUtils::EqualsNoCase(strProvidesButtonMaps, "true");
57
58       return std::unique_ptr<CPeripheralAddon>(new CPeripheralAddon(std::move(props), bProvidesJoysticks, bProvidesButtonMaps));
59     }
60
>>>     CID 1355587:  Performance inefficiencies  (PASS_BY_VALUE)
>>>     Passing parameter props of type "ADDON::AddonProps" (size 352 bytes) by value.
61     CPeripheralAddon::CPeripheralAddon(ADDON::AddonProps props, bool bProvidesJoysticks, bool bProvidesButtonMaps) :
62       CAddonDll<DllPeripheral, PeripheralAddon, PERIPHERAL_PROPERTIES>(std::move(props)),
63       m_apiVersion("0.0.0"),
64       m_bProvidesJoysticks(bProvidesJoysticks),
65       m_bProvidesButtonMaps(bProvidesButtonMaps)
66     {

** CID 1355585:  Control flow issues  (DEADCODE)
/xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamMultiSource.cpp: 112 in CInputStreamMultiSource::NextStream()()



________________________________________________________________________________________________________
*** CID 1355584:  Control flow issues  (DEADCODE)
/xbmc/addons/AddonBuilder.cpp: 140 in ADDON::CAddonBuilder::Build()()
134             return CAudioDecoder::FromExtension(std::move(m_props), m_extPoint);
135           else if (type == ADDON_INPUTSTREAM)
136             return CInputStream::FromExtension(std::move(m_props), m_extPoint);
137           else if (type == ADDON_SCREENSAVER)
138             return std::make_shared<CScreenSaver>(std::move(m_props));
139           else if (type == ADDON_PERIPHERALDLL)
>>>     CID 1355584:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "~unique_ptr" inside this statement: "<temporary>.~unique_ptr();".
140             return PERIPHERALS::CPeripheralAddon::FromExtension(std::move(m_props), m_extPoint);
141           break;
142         }
143         case ADDON_SKIN:
144           return CSkinInfo::FromExtension(std::move(m_props), m_extPoint);
145         case ADDON_RESOURCE_IMAGES:
```

OTHER CODE:

```
________________________________________________________________________________________________________
*** CID 1355585:  Control flow issues  (DEADCODE)
/xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamMultiSource.cpp: 112 in CInputStreamMultiSource::NextStream()()
106         if (next != NEXTSTREAM_NONE)
107           return next;
108       }
109       if (!eOF)
110         return NEXTSTREAM_RETRY;
111
>>>     CID 1355585:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "return CDVDInputStream::NEX...".
112       return NEXTSTREAM_NONE;
113     }
114
115     bool CInputStreamMultiSource::Open()
116     {


________________________________________________________________________________________________________
*** CID 1355583:  Error handling issues  (CHECKED_RETURN)
/xbmc/input/InputCodingTableBaiduPY.cpp: 50 in CInputCodingTableBaiduPY::Process()()
44
45     void CInputCodingTableBaiduPY::Process()
46     {
47       m_initialized = true;
48       while (!m_bStop) //Make sure we don't exit the thread
49       {
>>>     CID 1355583:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "AbortableWait" without checking return value (as is done elsewhere 8 out of 10 times).
50         AbortableWait(m_Event, -1); //Wait for work to appear
51         while (!m_bStop) //Process all queued work before going back to wait on the event
52         {
53           CSingleLock lock(m_CS);
54           if (m_work.empty())
55             break;


________________________________________________________________________________________________________
*** CID 1288117:  Error handling issues  (CHECKED_RETURN)
/xbmc/filesystem/RarFile.cpp: 62 in XFILE::CRarFileExtractThread::~CRarFileExtractThread()()
56       Create();
57     }
58
59     CRarFileExtractThread::~CRarFileExtractThread()
60     {
61       hQuit.Set();
>>>     CID 1288117:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "AbortableWait" without checking return value (as is done elsewhere 8 out of 10 times).
62       AbortableWait(hRestart);
63       StopThread();
64     }
65
66     void CRarFileExtractThread::Start(Archive* pArc, CommandData* pCmd, CmdExtract* pExtract, int iSize)
67     {



________________________________________________________________________________________________________
*** CID 1228822:  Error handling issues  (CHECKED_RETURN)
/xbmc/utils/ScraperParser.cpp: 458 in CScraperParser::Parse(const std::basic_string<char, std::char_traits<char>, std::allocator<char>>&, ADDON::CScraper *)()
452       if(pChildElement == NULL)
453       {
454         CLog::Log(LOGERROR,"%s: Could not find scraper function %s",__FUNCTION__,strTag.c_str());
455         return "";
456       }
457       int iResult = 1; // default to param 1
>>>     CID 1228822:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "QueryIntAttribute" without checking return value (as is done elsewhere 12 out of 15 times).
458       pChildElement->QueryIntAttribute("dest",&iResult);
459       TiXmlElement* pChildStart = FirstChildScraperElement(pChildElement);
460       m_scraper = scraper;
461       ParseNext(pChildStart);
462       std::string tmp = m_param[iResult-1];
463
```

garbear committed:
Remove unused #includes

garbear committed:
[peripherals] Fix Coverty issues (UNINIT_CTOR)

________________________________________________________________________________________________________
*** CID 1355594:  Uninitialized members  (UNINIT_CTOR)
/xbmc/peripherals/addons/AddonButtonMapping.cpp: 51 in PERIPHERALS::CAddonButtonMapping::CAddonButtonMapping(PERIPHERALS::CPeripheral *, JOYSTICK::IButtonMapper *)()
45         m_buttonMap = new CAddonButtonMap(peripheral, addon, mapper->ControllerID());
46         if (m_buttonMap->Load())
47           m_driverHandler = new CButtonMapping(mapper, m_buttonMap);
48         else
49           SAFE_DELETE(m_buttonMap);
50       }
>>>     CID 1355594:  Uninitialized members  (UNINIT_CTOR)
>>>     Non-static class member "m_buttonMap" is not initialized in this constructor nor in any functions that it calls.
51     }
52
53     CAddonButtonMapping::~CAddonButtonMapping(void)
54     {
55       delete m_driverHandler;
56       delete m_buttonMap;

________________________________________________________________________________________________________
*** CID 1355593:  Uninitialized members  (UNINIT_CTOR)
/xbmc/peripherals/addons/AddonInputHandling.cpp: 51 in PERIPHERALS::CAddonInputHandling::CAddonInputHandling(PERIPHERALS::CPeripheral *, JOYSTICK::IInputHandler *)()
45         m_buttonMap = new CAddonButtonMap(peripheral, addon, handler->ControllerID());
46         if (m_buttonMap->Load())
47           m_driverHandler = new CInputHandling(handler, m_buttonMap);
48         else
49           SAFE_DELETE(m_buttonMap);
50       }
>>>     CID 1355593:  Uninitialized members  (UNINIT_CTOR)
>>>     Non-static class member "m_buttonMap" is not initialized in this constructor nor in any functions that it calls.
51     }
52
53     CAddonInputHandling::~CAddonInputHandling(void)
54     {
55       delete m_driverHandler;
56       delete m_buttonMap;

garbear committed:
[controllers] Fix coverty issue (MISMATCHED_ITERATOR)

________________________________________________________________________________________________________
*** CID 1355586:  Control flow issues  (MISMATCHED_ITERATOR)
/xbmc/games/controllers/windows/GUIControllerList.cpp: 185 in GAME::CGUIControllerList::RefreshControllers()()
179
180       // Remove old controllers
181       for (ControllerVector::iterator it = m_controllers.begin(); it != m_controllers.end(); /* ++it */)
182       {
183         ControllerPtr& controller = *it;
184
>>>     CID 1355586:  Control flow issues  (MISMATCHED_ITERATOR)
>>>     Comparing "std::find_if(controllers.begin(), controllers.end(), GAME::CGUIControllerList::RefreshControllers()::[lambda(std::shared_ptr<GAME::CController> const &) (instance 2)](GAME::CGUIControllerList::RefreshControllers()::[lambda(std::shared_ptr<GAME::CController> const &) (instance 2)]({controller})))" from "controllers" to "newControllers.end()" from "newControllers".
185         if (std::find_if(controllers.begin(), controllers.end(),
186           [controller](const ControllerPtr& ctrl)
187           {
188             return ctrl->ID() == controller->ID();
189           }) == newControllers.end())
190         {

tamland commented on xbmc/peripherals/addons/AddonButtonMapping.cpp:
```diff
@@ -28,11 +28,12 @@ using namespace JOYSTICK;
 
 #ifndef SAFE_DELETE
-  #define SATE_DELETE(x)  do { delete (x); (x) = NULL; } while (0)
+  #define SATE_DELETE(x)  do { delete (x); (x) = nullptr; } while (0)
```
SA**T**E_DELETE? I believe you mean "UNSAFE_DELETE";) Please use smart pointers. The memory management in this class is broken. It will leak and double free.

garbear replied:
The memory management is flawed but not broken. Smart pointers should be used instead. will fix and PR when i get a chance

tamland replied:
That's fine but surely you should at least fix the typo ?

garbear committed:
[addons] Fix Coverty issue (DEADCODE)

The "Deadcode" it complains of is the "break;" statement below. All of the
case labels are accounted for in the if branching, so the final "break;" is
never hit.

________________________________________________________________________________________________________
*** CID 1355584:  Control flow issues  (DEADCODE)
/xbmc/addons/AddonBuilder.cpp: 140 in ADDON::CAddonBuilder::Build()()
134             return CAudioDecoder::FromExtension(std::move(m_props), m_extPoint);
135           else if (type == ADDON_INPUTSTREAM)
136             return CInputStream::FromExtension(std::move(m_props), m_extPoint);
137           else if (type == ADDON_SCREENSAVER)
138             return std::make_shared<CScreenSaver>(std::move(m_props));
139           else if (type == ADDON_PERIPHERALDLL)
>>>     CID 1355584:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "~unique_ptr" inside this statement: "<temporary>.~unique_ptr();".
140             return PERIPHERALS::CPeripheralAddon::FromExtension(std::move(m_props), m_extPoint);
141           break;
142         }
143         case ADDON_SKIN:
144           return CSkinInfo::FromExtension(std::move(m_props), m_extPoint);
145         case ADDON_RESOURCE_IMAGES:

garbear commented:
I've added a commit for another coverty issue, this one refactors some poor logic in the add-on builder. @MartijnKaijser fixes the coverty issue you pinged me on slack about.

jenkins build this please

tamland commented on xbmc/addons/AddonBuilder.cpp:
```diff
@@ -81,65 +116,26 @@ std::shared_ptr<IAddon> CAddonBuilder::Build()
-        return PERIPHERALS::CPeripheralAddon::FromExtension(std::move(m_props), m_extPoint);
-      break;
-    }
+      return PERIPHERALS::CPeripheralAddon::FromExtension(std::move(m_props), m_extPoint);
```
Can you please explain what the issue is and why it fixes it? The report from coverty makes no sense

garbear replied:
Correct, the coverty report made no sense. Martijn received an extended report with the words "must be true", which clued me into the defect. We have code like:

```c++
switch (foo) 
{
case A:
case B:
case C:
  if (A) ...
  else if (B) ...
  else if (C) ...
  break;
}
```
OFC, the `break` can never be hit, because C always "must be true".

This refactors the special casing to above the switch and simplifies the switch.

Paxxi commented:
I'll give a it a try to explain the pass by value warnings coverity gives, let's start with a call stack for one scenario
```
std::shared_ptr<IAddon> CAddonBuilder::Build()
return CPluginSource::FromExtension(std::move(m_props), m_extPoint);
return std::unique_ptr<CPluginSource>(new CPluginSource(std::move(props), provides));
```
What will happen here is that the compiler will not move m_props for two reasons.
1. FromExtension does not take an rvalue (in itself not an issue)
2. m_props is a member variable and moving from it would leave it either in it's default state or in an unkown state depending on how the move operator is implemented.(not sure about the default now but I believe it's using std::swap)

So far we have one copy of m_props into FromExtension which coverity complains about being expensive as m_props is fairly large.
Next step will be passing props into CPluginSource() which also does not take an rvalue but in this case the compiler should figure out that props won't be used anymore and move it as specified avoiding a second copy.(Not sure if it's guaranteed or compiler specific, haven't read the standard).

So in the end we know we get one copy operation and in worst case scenario we may get two copies for each call to `CAddonBuilder::Build()`

### Lessons about std::move ###
1. std::move does not guarantee that an object will be moved, the only thing it does is change an lvalue into an rvalue and then it's up to the compiler and the standard to decide if the rvalue should be moved or copied which can be tricky to reason about
2. _NEVER_ ever use std::move from a member variable unless it's in a move constructor or move operator as it will yank the data out of that variable and leave it in either a default state or a broken state depending on the move operator for that type.

#### What on earth does rvalue and lvalue mean? ####
I can't explain this in exact detail as with everything there are exceptions but the rule of thumb that works is
* lvalue is any variable that has a name, e.g a member variable, local variable or an argument to a method. lvalue is short for left-hand-value basically meaning to the left of an assignment operator.
* rvalue is an unbound variable such as a return value from a method, it's allocated in memory somewhere but it has no explicit name. It will become an lvalue once it's assigned to a named variable. rvalue is short for right-hand-value basically meaning to the right of an assignment operator.

This comment ended up a bit long but I thought I'd take the opportunity to try and spread some info about std::move :smile:

garbear commented:
Updated the `UNINIT_CTOR` commit to use smart pointers per discussion with @tamland

garbear commented:
jenkins build this please

garbear commented:
Merging. In the future, I'll address the PASS_BY_VALUE cases via advice from @paxxi

tamland commented:
>What will happen here is that the compiler will not move m_props for two reasons.

Do you have references for that? To me this is not obvious.


Please keep in mind that converity does _static analysis_ and that many 'issues' are nothing more than heuristics and guesses based on common c++ mistakes. Pretty sure pass by value is is just a simple check that looks at function signature and an estimated size and print these warnings. It does so even when a copy is required, and even when it's never actually copied. Don't automatically assume it's anything more than that.
